File: tree/binary_search_tree/108_convert_sorted_array_to_binary_search_tree.cc
#include <vector>

#include "tree_node.h"

// 2024/01/09: 完全不看答案自己实现

TreeNode* SortedArrayToBST(const std::vector<int>& nums) {
  return SortedArrayToBST(nums, 0, nums.size() - 1);
}

// [left, right]
TreeNode* SortedArrayToBST(const std::vector<int>& nums, int left, int right) {
  if (left > right) {
    return nullptr;
  }
  int mid = left + (right - left) / 2;
  TreeNode* root = new TreeNode(nums[mid]);
  root->left = SortedArrayToBST(nums, left, mid - 1);
  root->right = SortedArrayToBST(nums, mid + 1, right);
  return root;
}

File: tree/binary_search_tree/108_convert_sorted_array_to_binary_search_tree.py
from typing import List, Optional
from tree.tree_node import TreeNode, print_tree_node


# 2026/05/18: 自己写出答案
def sorted_array_to_bst(nums: List[int]) -> Optional[TreeNode]:
    if len(nums) == 0:
        return None
    counts = len(nums)
    mid = counts // 2
    node = TreeNode(nums[mid])
    node.left = sorted_array_to_bst(nums[:mid])
    node.right = sorted_array_to_bst(nums[mid + 1 :])
    return node


nums = [-10, -3, 0, 5, 9]
print_tree_node(sorted_array_to_bst(nums))
